src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs

```diff
@@ -526,4 +515,42 @@ private static List<TestResult> UpdateResultsWithParentInfo(
 
         return updatedResults;
     }
+
+    private bool TryExecuteITestDataSource(List<TestResult> results)
+    {
+        if (_test.DataType != DynamicDataType.ITestDataSource)
+        {
+            return false;
+        }
+
+        UTF.ITestDataSource? dataSource;
+        object?[]? data;
+        if (_test.SerializedData?.Length == 3)
+        {
+            if (!Enum.TryParse(_test.SerializedData[0], out TestDataSourceUnfoldingStrategy _)
+                || !int.TryParse(_test.SerializedData[1], out int dataSourceIndex)
+                || !int.TryParse(_test.SerializedData[2], out int dataIndex))
+            {
+                throw ApplicationStateGuard.Unreachable();
+            }
+
+            dataSource = _testMethodInfo.GetAttributes<Attribute>(false)?.OfType<UTF.ITestDataSource>().Skip(dataSourceIndex).FirstOrDefault();
```

Comment:
Runtime team just told me not to rely on guaranteed reflection order. I'll try to see what other option we have. Given that this is currently always working fine, I'd vote for adding a comment and moving on with the PR.